### ISSUE TITLE
SDK is overriding the Activity.Current.TraceId even with a custom DistributedContextPropagator

### ISSUE THREAD

**Issue by timmydo** [open-telemetry/opentelemetry-dotnet#4310]
# Bug Report

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):
```
    <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.14" />
    <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc9.14" />
```

* Sample Repro

```
namespace AspNetInstrumentationHeader
{
    internal class Program
    {
        static void Main(string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);
            builder.Services.AddSingleton<DistributedContextPropagator>(new MyDistributedContextPropagator());
            var tracer = OpenTelemetry.Sdk.CreateTracerProviderBuilder()
            .SetSampler(new AlwaysOnSampler())
            .AddAspNetCoreInstrumentation()
            .Build();

            var app = builder.Build();
            app.Map("/", () =>
            {
                return $"{Activity.Current.TraceId}";
            });

            app.Run();
        }

 
        private class MyDistributedContextPropagator : DistributedContextPropagator
        {
            private readonly DistributedContextPropagator defaultImpl = DistributedContextPropagator.Current;

            public override IReadOnlyCollection<string> Fields => this.defaultImpl.Fields;
 
            public override IEnumerable<KeyValuePair<string, string?>>? ExtractBaggage(object? carrier, PropagatorGetterCallback? getter)
            {
                return this.defaultImpl.ExtractBaggage(carrier, getter);
            }

            public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceId, out string? traceState)
            {
                traceId = Guid.Empty.ToString("n");
                traceState = string.Empty;
            }
 

            public override void Inject(Activity? activity, object? carrier, PropagatorSetterCallback? setter)
            {
                this.defaultImpl.Inject(activity, carrier, setter);
            }
        }
    }
}

 
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

 
  <ItemGroup>
    <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.14" />
    <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc9.14" />
  </ItemGroup>
</Project>


curl -H "traceparent:00-64125a8060b946c7baa34ba37cba3b4a-d4e8b89a634d5b06-01" http://localhost:5000/
64125a8060b946c7baa34ba37cba3b4a
```

## Symptom

Seems `.AddAspNetCoreInstrumentation()` overrides traceid from our DistributedContextPropagator

**What is the expected behavior?**

Expected to get the trace ID from our DistributedContextPropagator

**What is the actual behavior?**

Guid came from traceparent header.

**Comment by reyang** [comment]
@vishweshbankwar could you look into this?

**Comment by cijothomas** [comment]
AspNetCoreInstrumentation does not support DistributedContextPropagatos. It only supports the Propagators model from OpenTelemetry project, so this is just expected behavior.

You'll need to override the OTEL propagators to achieve the desired behaviour (https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Api#context-propagation). It defaults to a compose-propagator that has W3CTraceContextPropagator inside it, which is the component causing the issue in your case.

This is how you need to override the default:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Sdk.cs#L54

(This part is not documented very well, so this is going to require some manual exploration. We can help as needed)

To summarize:
1. Asp.Net Core by default uses traceid from incoming traceparent header. (this is irrespective of Otel or not)
2. You can override Asp.Net Core's default behavior by using the custom DistributedContextPropagator, as you have done already.
3. OTel.Instrumentation.AspNetCore (which you have enabled), uses OTels default propagator, which will also use traceid from incoming traceparent. 
4. To override 3, you need to use Sdk.SetDefaultTextMapPropagator to point to a custom propagator which should ignore incoming traceparent.

(The fact that OTel Propagators and the Propagators from .NET Runtime are different entities today is the cause of confusion, This is something we need to solve. @vishweshbankwar has some explorations to make it happen)

**Comment by timmydo** [comment]
so is this the right way? add this before app.run:
```
            OpenTelemetry.Sdk.SetDefaultTextMapPropagator(new CompositeTextMapPropagator(new TextMapPropagator[]
            {
                new MyTraceContextPropagator(),
            }));
```

defined as:
```
        private class MyTraceContextPropagator : TextMapPropagator
        {
            public override ISet<string> Fields => new HashSet<string>();

            public override PropagationContext Extract<T>(PropagationContext context, T carrier, Func<T, string, IEnumerable<string>> getter)
            {
                var tid = ActivityTraceId.CreateFromBytes(Guid.Empty.ToByteArray());
                var ctx = new ActivityContext(tid, ActivitySpanId.CreateRandom(), context.ActivityContext.TraceFlags, context.ActivityContext.TraceState, isRemote: true);
                return new PropagationContext(ctx, context.Baggage);
            }

            public override void Inject<T>(PropagationContext context, T carrier, Action<T, string, string> setter)
            {
                return;
            }
        }
```

**Comment by cijothomas** [comment]
^ Yes but be aware that the Propagator affect any other instrumentation as well in your app. (there is no per-instrumentation library propagator support).

Also it looks like the inject is not doing actual injection, make sure you are okay with that. (if only using httpclient, then it should be fine as httpclient natively injects context.)

Other note: You'll miss OTel baggage, as the BaggagePropagator is not added to the composite one. If you don't need it, it is okay.
